diesel/src/query_builder/collected_query.rs

```diff
@@ -0,0 +1,48 @@
+use super::{AstPass, MoveableBindCollector, Query, QueryFragment, QueryId};
+use crate::backend::{Backend, DieselReserveSpecialization};
+use crate::result::QueryResult;
+use crate::sql_types::Untyped;
+
+#[derive(Debug)]
+#[must_use = "Queries are only executed when calling `load`, `get_result` or similar."]
+/// A SQL query variant with already collected bind data which can be moved
+pub struct CollectedQuery<T> {
+    sql: String,
+    safe_to_cache_prepared: bool,
+    bind_data: T,
+}
+
+impl<T> CollectedQuery<T> {
+    /// Builds a [CollectedQuery] with movable bind data
+    pub fn new(sql: String, safe_to_cache_prepared: bool, bind_data: T) -> Self {
+        Self {
+            sql,
+            safe_to_cache_prepared,
+            bind_data,
+        }
+    }
+}
+
+impl<DB, T> QueryFragment<DB> for CollectedQuery<T>
+where
+    DB: Backend + DieselReserveSpecialization,
+    for<'a> <DB as Backend>::BindCollector<'a>: MoveableBindCollector<DB, BindData = T>,
+{
+    fn walk_ast<'b>(&'b self, mut pass: AstPass<'_, 'b, DB>) -> QueryResult<()> {
+        if !self.safe_to_cache_prepared {
+            pass.unsafe_to_cache_prepared();
+        }
+        pass.push_sql(&self.sql);
+        pass.push_bind_collector_data::<T>(&self.bind_data)
+    }
+}
+
+impl<T> QueryId for CollectedQuery<T> {
+    type QueryId = ();
+
+    const HAS_STATIC_QUERY_ID: bool = false;
+}
+
+impl<T> Query for CollectedQuery<T> {
+    type SqlType = Untyped; // FIXME set proper type of underlying original query
```

Comment:
We should be fine to just have a `p: PhantomData<ST>` field on the `CollectedQuery` struct which uses an additional generic type `ST`, which corresponds tho the sql type of the original query.

Reply:
I am trying to figure out a solution here ([diesel branch `sql](https://github.com/WattSense/diesel/tree/sql_type_for_collected_query and https://github.com/WattSense/diesel_async/tree/collected_query_sqltype) but it needs bound on HasSqlType<T>.

Reply:
Right, that won't work. In that case: I think using `Untyped` there is fine, as it doesn't allow anything bad anyway. It just restricts that if you manage to construct that type you are responsible for deserializing results on your own (as we do in diesel-async) or that you use `QueryableByName`. 

So just removing that fixme should resolve this issue.

Reply:
Thanks for the explanation, I was wondering how that works since it is possible to get structs from the load result.